// firebase-functions/lib/esm/v2/providers/alerts/index.mjs
import { __export } from "../../../_virtual/rolldown_runtime.mjs";
import { convertAlertAndApp, eventType, getEndpointAnnotation, getOptsAndAlertTypeAndApp, onAlertPublished } from "./alerts.mjs";
import { appDistribution_exports } from "./appDistribution.mjs";
import { billing_exports } from "./billing.mjs";
import { crashlytics_exports } from "./crashlytics.mjs";
import { performance_exports } from "./performance.mjs";

//#region src/v2/providers/alerts/index.ts
var alerts_exports = /* @__PURE__ */ __export({
	appDistribution: () => appDistribution_exports,
	billing: () => billing_exports,
	convertAlertAndApp: () => convertAlertAndApp,
	crashlytics: () => crashlytics_exports,
	eventType: () => eventType,
	getEndpointAnnotation: () => getEndpointAnnotation,
	getOptsAndAlertTypeAndApp: () => getOptsAndAlertTypeAndApp,
	onAlertPublished: () => onAlertPublished,
	performance: () => performance_exports
});

//#endregion
export { alerts_exports, appDistribution_exports as appDistribution, billing_exports as billing, convertAlertAndApp, crashlytics_exports as crashlytics, eventType, getEndpointAnnotation, getOptsAndAlertTypeAndApp, onAlertPublished, performance_exports as performance };
