// firebase-functions/lib/esm/v1/index.mjs
import { logger_exports } from "../logger/index.mjs";
import { params_exports } from "../params/index.mjs";
import { RESET_VALUE } from "../common/options.mjs";
import { firebaseConfig } from "../common/config.mjs";
import { setApp } from "../common/app.mjs";
import { Change } from "../common/change.mjs";
import { onInit } from "../common/onInit.mjs";
import { DEFAULT_FAILURE_POLICY, INGRESS_SETTINGS_OPTIONS, MAX_NUMBER_USER_LABELS, MAX_TIMEOUT_SECONDS, MIN_TIMEOUT_SECONDS, SUPPORTED_REGIONS, VALID_MEMORY_OPTIONS, VPC_EGRESS_SETTINGS_OPTIONS } from "./function-configuration.mjs";
import { makeCloudFunction, optionsToEndpoint, optionsToTrigger } from "./cloud-functions.mjs";
import { config } from "./config.mjs";
import { analytics_exports } from "./providers/analytics.mjs";
import { auth_exports } from "./providers/auth.mjs";
import { database_exports } from "./providers/database.mjs";
import { firestore_exports } from "./providers/firestore.mjs";
import { https_exports } from "./providers/https.mjs";
import { pubsub_exports } from "./providers/pubsub.mjs";
import { remoteConfig_exports } from "./providers/remoteConfig.mjs";
import { storage_exports } from "./providers/storage.mjs";
import { tasks_exports } from "./providers/tasks.mjs";
import { testLab_exports } from "./providers/testLab.mjs";
import { FunctionBuilder, region, runWith } from "./function-builder.mjs";

//#region src/v1/index.ts
const app = { setEmulatedAdminApp: setApp };

//#endregion
export { Change, DEFAULT_FAILURE_POLICY, FunctionBuilder, INGRESS_SETTINGS_OPTIONS, MAX_NUMBER_USER_LABELS, MAX_TIMEOUT_SECONDS, MIN_TIMEOUT_SECONDS, RESET_VALUE, SUPPORTED_REGIONS, VALID_MEMORY_OPTIONS, VPC_EGRESS_SETTINGS_OPTIONS, analytics_exports as analytics, app, auth_exports as auth, config, database_exports as database, firebaseConfig, firestore_exports as firestore, https_exports as https, logger_exports as logger, makeCloudFunction, onInit, optionsToEndpoint, optionsToTrigger, params_exports as params, pubsub_exports as pubsub, region, remoteConfig_exports as remoteConfig, runWith, storage_exports as storage, tasks_exports as tasks, testLab_exports as testLab };
